recipes/full_finetune_distributed.py

```diff
@@ -199,21 +204,101 @@ def __init__(self, cfg: DictConfig) -> None:
         self.total_epochs = cfg.epochs
         self.max_steps_per_epoch = cfg.max_steps_per_epoch
         self.global_step = 0
+        self._dcp_checkpointer = None
+
+    def get_dcp_checkpointer(self):
+        if not self._dcp_checkpointer:
+            self._dcp_checkpointer = DistributedCheckpointer(
+                checkpoint_dir=self._checkpointer._checkpoint_dir,
+                output_dir=self._checkpointer._output_dir,
+            )
+
+        return self._dcp_checkpointer
 
     def load_checkpoint(self, cfg_checkpointer: DictConfig) -> Dict[str, Any]:
         """
-        Extract the checkpoint state from file and validate. If resume_from_checkpoint
-        is True, this also includes the recipe state.
+        Extract the checkpoint state from file and validate.
+        If resume_from_checkpoint is True, this also includes the recipe state.
         """
+        # If async checkpointing is enabled, set the should_load_recipe_state as false for the checkpointer.
+        # We will use DistributedCheckpointer instead to load the intermediate checkpoints which includes the
+        # recipe state as well.
+        should_load_recipe_state: bool = (
+            False if self._enable_async_checkpointing else self._resume_from_checkpoint
+        )
+
         self._checkpointer = config.instantiate(
             cfg_checkpointer,
-            resume_from_checkpoint=self._resume_from_checkpoint,
+            should_load_recipe_state=should_load_recipe_state,
         )
-        checkpoint_dict = self._checkpointer.load_checkpoint()
 
-        if self._resume_from_checkpoint:
+        return self._checkpointer.load_checkpoint()
```

Comment:
These two functions are doing the same, but their signature is very different. Async also depends on the other, instead of config args.

for dcp, we are doing
```
dcp_checkpointer = self.get_dcp_checkpointer()
state_dict = dcp_checkpointer.load_checkpoint(state_dict)
```

for sync ckpt, we are doing in setup:
```
# Load the base model
checkpoint_dict = self.load_checkpoint(cfg_checkpointer=cfg.checkpointer)
```

I understand that you want to avoid a major refactoring. However, in my opinion, these functions should just be:

```
def setup_async_ckpt(self, cfg_checkpointer):
	return DistributedCheckpointer(
                checkpoint_dir=cfg_checkpointer.checkpoint_dir,
                output_dir=cfg_checkpointer.output_dir,
            )
	
def setup_sync_ckpt(self, cfg_checkpointer):
	return config.instantiate(
            cfg_checkpointer,
            resume_from_checkpoint=self._resume_from_checkpoint,
            should_load_recipe_state=should_load_recipe_state,
        )
```

and call:
```
self.dcp_checkpointer = setup_async_ckpt(cfg_checkpointer)
state_dict = self.dcp_checkpointer .load_checkpoint(state_dict)
```

Reply:
The selection of correct Checkpointer has been abstracted away into a CheckpointClient which will encapsulate checkpointing logic for all recipes. This reduces the checkpointing code footprint in the recipes and will make it much easier to introduce async checkpoints for other recipes.